src/util/bson.hpp

```diff
@@ -0,0 +1,1218 @@
+                    auto regex = std::get<RegularExpression>(m_marks.top().back().second);
+                    m_marks.top().pop_back();
+                    m_marks.top().push_back({instruction.key, RegularExpression(regex.pattern(), val)});
+                    // pop vestigal regex instruction
+                    m_instructions.pop();
+                    m_instructions.push({state_t::skip});
+                    m_instructions.push({state_t::skip});
+                } else {
+                    m_marks.top().push_back({instruction.key, RegularExpression("", val)});
+                }
+
+                break;
+            case state_t::binary_sub_type:
+                // if we have already pushed a binary type
+                if (m_marks.top().size() && m_marks.top().back().first == instruction.key) {
+                    // we will ignore the subtype for now
+                    // pop vestigal binary instruction
+                    m_instructions.pop();
+                    m_instructions.push({state_t::skip});
+                    m_instructions.push({state_t::skip});
+                } else {
+                    // we will ignore the subtype for now
+                    m_marks.top().push_back({instruction.key, BinaryData()});
+                }
+
+                break;
+            case state_t::binary_base64: {
+                char* binary_val = (char *)malloc(sizeof(char) * val.size());
+                BinaryData binary_data;
+                if (val.size()) {
+                    const char* copy = strcpy(binary_val, val.data());
+                    binary_data = BinaryData(copy, strlen(copy) + 1);
+                }
+                // if we have already pushed a binary type
+                if (m_marks.top().size() && m_marks.top().back().first == instruction.key) {
+                    m_marks.top().pop_back();
+                    m_marks.top().push_back({instruction.key, binary_data});
+
+                    // pop vestigal binary instruction
+                    m_instructions.pop();
+                    m_instructions.push({state_t::skip});
+                    m_instructions.push({state_t::skip});
+                } else {
+                    // we will ignore the subtype for now
+                    m_marks.top().push_back({instruction.key, binary_data});
+                }
+
+                break;
+            }
+            case state_t::json_key: {
+                char* v = (char *)malloc(sizeof(char) * val.size());
+                m_marks.top().push_back({instruction.key, strcpy(v, val.data())});
+                break;
+            }
+            default:
+                check_state(instruction.type, state_t::json_key);
+                break;
+        }
+        return true;
+    }
+
+    /*!
+    @brief an object key was read
+    @param[in] val  object key
+    @return whether parsing should proceed
+    @note It is safe to move the passed string.
+    */
```

Comment:
I think we prefer single line `//` comments in this repository for documentation, IIRC the reasoning is that it is harder to turn things off experimentally or during development by using multiline comments like this because these doc comments will close things off.